### PR TITLE
[metrics] IoU Metric

### LINKED ISSUE #1299: [Metrics] IOU
## 🚀 Feature
Implement (differentiable) IOU

### PR DESCRIPTION
**Pull request by j-dsouza**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Adds feature, fixes #1299 

WIP:
- [x] IoU metric class to be added
- [x] Documentation to be added

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/metrics/functional/classification.py [hunk 1]
```diff
@@ -24,10 +24,8 @@ def to_categorical(tensor: torch.Tensor, argmax_dim: int = 1) -> torch.Tensor:
 def get_num_classes(pred: torch.Tensor, target: torch.Tensor,
                     num_classes: Optional[int]) -> int:
     if num_classes is None:
-        if pred.ndim > target.ndim:
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:27]: I don't think you can discard these...

File: tests/metrics/functional/test_classification.py [hunk 2]
```diff
@@ -294,3 +297,28 @@ def test_dice_score():
 
 # example data taken from
 # https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/tests/test_ranking.py
+
+
+def test_iou():
+    # Generate dataset
+    target = (torch.arange(120) % 3).view(-1, 1)
```
Comment (justusschock) [inline tests/metrics/functional/test_classification.py:304]: can you also use pytest parametrization? That way we can easily add additional testcases :)

File: pytorch_lightning/metrics/functional/classification.py [hunk 3]
```diff
@@ -397,3 +398,34 @@ def dice_score(pred: torch.Tensor, target: torch.Tensor, bg: bool = False,
...
+
+    Returns:
+        IoU score : Tensor containing single value if reduction is
+            'elementwise_mean', or number of classes if reduction is 'none'
+    """
+    cm = confusion_matrix(pred, target, normalize=True)
+    if remove_bg:
+        cm = cm[1:, 1:]
+    tp = torch.diag(cm)
+    fp = cm.sum(dim=0) - tp
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:428]: we basically also have a function to calculate these values without calculating a whole cm

File: pytorch_lightning/metrics/functional/classification.py (hunk 1 above)
Comment (j-dsouza) [reply pytorch_lightning/metrics/functional/classification.py]: Good point. Added back in

File: pytorch_lightning/metrics/functional/classification.py (hunk 3 above)
Comment (j-dsouza) [reply pytorch_lightning/metrics/functional/classification.py]: Updated - Missed these first time, sorry!

File: tests/metrics/functional/test_classification.py (hunk 2 above)
Comment (j-dsouza) [reply tests/metrics/functional/test_classification.py]: Updated

**Comment by Borda** [comment]
@j-dsouza @justusschock how is it going here? maybe set `master` as a target? :rabbit:

**Comment by justusschock** [comment]
This should be ready, once #1488 is done. We cannot yet set master as target, since this depends on some functions of #1488

**Comment by justusschock** [comment]
@Borda why close this? This is not included by #1488 but depends on it

**Comment by Borda** [comment]
it was some automation from GH, can we reopen it?

**Comment by Borda** [comment]
@j-dsouza ready to review?

**Comment by williamFalcon** [comment]
let's get this in

**Comment by j-dsouza** [comment]
Sorry for the slow reply @justusschock and @Borda - I've been away for a couple of days

Should have time to add the module inferface for tests tomorrow, so will let you know when that's done and then should be good to review :)

**Comment by j-dsouza** [comment]
@justusschock - This should be ready to go now :)

**Comment by Borda** [comment]
the docs shall be fixed in #2227

File: tests/metrics/functional/test_classification.py [hunk 4]
```diff
@@ -366,5 +367,25 @@ def test_dice_score(pred, target, expected):
...
+                 True, 'none', False, torch.Tensor([0.5, 0.5, 0.5])),
+    pytest.param((torch.arange(120) % 3).view(-1, 1), (torch.arange(120) % 3).view(-1, 1),
+                 True, 'elementwise_mean', False, torch.Tensor([0.5])),
+    pytest.param((torch.arange(120) % 3).view(-1, 1), (torch.arange(120) % 3).view(-1, 1),
+                 True, 'none', True, torch.Tensor([0.5, 0.5])),
+])
+def test_iou(target, pred, half_ones, reduction, remove_bg, expected):
+    if half_ones:
+        pred[:60] = 1
+    assert torch.all(torch.eq(iou(pred, target, remove_bg=remove_bg, reduction=reduction), expected))
```
Comment (Borda) [inline tests/metrics/functional/test_classification.py:387]:
```suggestion
@pytest.mark.parametrize(['half_ones', 'reduction', 'remove_bg', 'expected'], [
    pytest.param(False, 'none', False, torch.Tensor([1, 1, 1])),
    pytest.param(False, 'elementwise_mean', False, torch.Tensor([1])),
    pytest.param(False, 'none', True, torch.Tensor([1, 1])),
    pytest.param(True, 'none', False, torch.Tensor([0.5, 0.5, 0.5])),
    pytest.param(True, 'elementwise_mean', False, torch.Tensor([0.5])),
    pytest.param(True, 'none', True, torch.Tensor([0.5, 0.5])),
])
def test_iou(half_ones, reduction, remove_bg, expected):
    pred = (torch.arange(120) % 3).view(-1, 1)
    target = (torch.arange(120) % 3).view(-1, 1)
    if half_ones:
        pred[:60] = 1
    iou_val = iou(pred, target, remove_bg=remove_bg, reduction=reduction)
    assert torch.allclose(iou_val, expected, atol=1e-9)
```

File: tests/metrics/test_classification.py [hunk 5]
```diff
@@ -205,3 +206,14 @@ def test_dice_coefficient(include_background):
...
+
+@pytest.mark.parametrize('remove_bg', [True, False])
+def test_iou(remove_bg):
+    iou = IoU(remove_bg=remove_bg)
+    assert iou.name == 'iou'
+
+    score = iou(torch.randint(0, 1, (10, 25, 25)),
+                torch.randint(0, 1, (10, 25, 25)))
+
+    assert isinstance(score, torch.Tensor)
```
Comment (Borda) [inline tests/metrics/test_classification.py:219]: add test for result value

File: pytorch_lightning/metrics/classification.py [hunk 6]
```diff
@@ -770,3 +771,47 @@ def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
+                 remove_bg: bool = False,
+                 reduction: str = 'elementwise_mean'):
+        """
+        Args:
+            remove_bg: Flag to state whether a background class has been included
+                within input parameters. If true, will remove background class. If
+                false, return IoU over all classes.
+                Assumes that background is '0' class in input tensor
+            reduction: a method for reducing IoU over labels (default: takes the mean)
+                Available reduction methods:
```
Comment (Borda) [inline pytorch_lightning/metrics/classification.py:803]:
```suggestion
                Available reduction methods:

```

**Comment by Borda** [comment]
@j-dsouza great work!

**Pull request merged**